controller/api/destination/server.go

```diff
@@ -383,11 +427,38 @@ func getSvcID(k8sAPI *k8s.API, clusterIP string, log *logging.Entry) (*watcher.S
 	return service, nil
 }
 
-// getPod returns a pod that maps to the given IP address. The pod can either
+// getPodByHostname returns a pod that maps to the given hostname (or an
+// instanceID). The hostname is generally the prefix of the pod's DNS name;
+// since it may be arbitrary we need to look at the corresponding service's
+// Endpoints object to see whether the hostname matches a pod.
+func getPodByHostname(k8sAPI *k8s.API, hostname string, svcID watcher.ServiceID) (*corev1.Pod, error) {
+	ep, err := k8sAPI.Endpoint().Lister().Endpoints(svcID.Namespace).Get(svcID.Name)
```

Comment:
It may worth adding an indexer for endpoints by service name in `InitializeIndexers`. I'm not sure it's a blocker for this, but this could be a hot path and we may see issues similar to the past when we've used listers.

Also, the statefulsets change has an endpoints lister as well (it may have been for this though). If they are different, then both changes could benefit from an indexer.

I'd be fine coming back to this as a follow-up but just want to get conversation started on it.

Reply:
@kleimkuhler I spent some time digging into this today. I really like the idea of building an indexer specifically for Endpoints but I'm a bit curious as to how we can do it without having potential clashes. Indexing by service name in this case would mean indexing by the name of the Endpoints object (since they share the same name as the service). Without a namespace I think we can get into clashes (since the same service name may be used across namespaces).

I think resolving the clash after might have the same efficiency as a Lister? From what I understand, the default indexer is based on namespace name -- I assume after we find objects with a matching namespace they're filtered based on the object kind and name (as is the case with `.Endpoints(ns).Get(name)`. Wdyt about building an index based on the entire ID? e.g `serviceNamespace/serviceName` -- we always have access to both and can make use of it in the statefulset PR too. I'm also open to other approaches here, I think we can get it done as part of this change.

Reply:
I'm unclear on why we'd use an indexer here?  We have the Endpoints name and namespace and are fetching it directly.  What would indexing accomplish?

Reply:
@adleong I had a chat with @kleimkuhler off GitHub about this. We arrived to the conclusion that we wouldn't need to have an indexer. I think at first we had some misconceptions on how client-go works underneath but we managed to work them out in our conversation; these misconceptions were mostly around the use of informers and whether an informer's Lister makes full use of the  cache/object store that it keeps in memory -- it does.